### PR TITLE
refactor(providers): tighten contract — classification, sequence lock, idempotency, unsub registry

### PR DESCRIPTION
**Pull request by raman325**
## Proposed change

Five related provider-layer cleanups, mostly one per commit. The first four were the original scope; the fifth was surfaced by review.

### 1. Failure classification post-#1184 (`ec0eb15a`)

After #1184 routes service-call failures to `LockOperationFailed` and breaks per-slot on that, the wrong exception classification has user-visible consequences: a transient communication failure misclassified as `LockOperationFailed` now breaks the slot instead of triggering reconnect.

- **ZHA**: `cluster.set_pin_code` / `cluster.clear_pin_code` bare-`Exception` catch now raises `LockDisconnected` instead of `LockOperationFailed`. Generic exceptions from a Zigbee cluster call are overwhelmingly comms/timing issues, not operational rejections. The non-zero-status branch still raises `CodeRejectedError`.
- **Zigbee2MQTT**: split MQTT publish errors — `OSError` → `LockDisconnected` (transient/connectivity), `HomeAssistantError` → `LockOperationFailed` (operational).
- **Matter** is left unchanged for now: Matter helpers raise `HomeAssistantError` for both transient comms failure and operational rejection, with no finer-grained type available. Currently all map to `LockDisconnected`. Flagged as an out-of-scope item for the eventual Matter data-model rewrite.

### 2. Sequence lock for multi-step provider ops (`99d2861b`)

Schlage and Akuvox do read-modify-write sequences (`get-codes → delete → add` for renames; `get → delete → set` for `async_set_usercode`). Only the leaf service calls go through `_execute_rate_limited`, which means concurrent callers can interleave and produce duplicate names / "already exists" loops (already noted as eventual-consistency comments in the code).

New base context manager `_serialize_sequence()` returns a separate `_sequence_lock` distinct from `_aio_lock`. Leaf calls inside a sequence still go through `_execute_rate_limited` against `_aio_lock` without deadlocking. Schlage and Akuvox wrap their multi-step paths.

### 3. Idempotent setup tag pass (`99d2861b`)

Schlage and Akuvox auto-tag unmanaged codes at setup. The base contract says `async_setup` must be idempotent; running the tag pass again on reconnect against a possibly-drifted device list produces double-tag / rename storms.

Adds `_tagged_once: bool` on both providers; the setup tag pass short-circuits if already True. Hard refresh bypasses the gate by calling the inner `_async_run_tag_pass` directly (the deliberate-reset case).

### 4. Push-unsub registry on the base (`aaba229f`, `38286f1f`)

Each provider previously invented its own unsub field — `_event_unsub` (Matter), `_cluster_listener_unsub` (ZHA), `_unsubscribe` (Zigbee2MQTT), `_value_update_unsub` (Z-Wave JS). Lifted to a shared `_push_unsubs: list[Callable[[], None]]` registry on the base with `_register_push_unsub` / `_clear_push_unsubs` helpers. Each unsub is invoked in try/except so one provider's broken teardown does not strand the others.

The `_clear_push_unsubs` follow-up (`b2118f22`) snapshots the registry before invoking so an unsub that re-registers won't corrupt iteration, and tightens the docstring to call out that Home Assistant event-bus listeners (Z-Wave JS's `_listeners`, lifecycle = setup/unload) must NOT be registered here.

Z-Wave JS's separate `_listeners` list (HA event-bus listeners, setup/unload lifecycle) is intentionally kept distinct from push-subscription teardown.

### 5. Tag-pass disconnect propagation so the idempotency guard does its job (`c06797c9`)

Surfaced by review: the original `_async_run_tag_pass` in Schlage/Akuvox caught `(LockDisconnected, LockOperationFailed)` as a single class and logged-and-continued. A hard-disconnect storm during the first setup pass would leave every per-code attempt logged but the function would return normally, `async_setup` would set `_tagged_once = True`, and no reconnect would ever retry the pass — defeating the guard's purpose.

Both providers now distinguish `LockDisconnected` from `LockOperationFailed` in the per-code (per-user) catches: operation failures stay per-code (do not gate idempotency), but a disconnect anywhere in the pass flips a `saw_disconnect` flag that re-raises `LockDisconnected` at the end. The re-raise propagates through `_async_tag_unmanaged_codes` / `_async_tag_unmanaged_users` into `async_setup` so the `_tagged_once = True` assignment is skipped and `_async_on_integration_loaded` retries on reconnect. Also fixes the latent `except LockDisconnected, LockOperationFailed:` style at `akuvox.py:249` (Py3 tuple-expression form — functional but stylistically broken).

## Out-of-scope findings worth flagging

- Matter's coarse `HomeAssistantError` → `LockDisconnected` mapping → PR F.
- `except E1, E2:` Py3 tuple-expression style appears in 4 remaining places (`_base.py:508`, `_util.py:48`, `zwave_js.py:157`, `zha.py:399, 444`) — functional, cosmetic only.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

Part of the resilience refactor chain. Depends on #1187 (lifecycle hygiene).

- This PR fixes or closes issue: fixes #
- This PR is related to issue:

### CODE REVIEW

File: custom_components/lock_code_manager/providers/_base.py [hunk 1]
```diff
@@ -239,6 +249,45 @@ async def _execute_rate_limited(
...
+        for unsub in self._push_unsubs:
+            try:
+                unsub()
+            except Exception as err:
+                LOGGER.warning(
+                    "Lock %s: push unsubscribe raised, continuing teardown: %s",
+                    self.lock.entity_id,
+                    err,
+                )
+        self._push_unsubs.clear()
```
[the remark this answers is not in the document]
Comment (raman325) [reply custom_components/lock_code_manager/providers/_base.py]: Fixed in b2118f22. Snapshot `_push_unsubs` and clear the registry before invoking from the snapshot. Now re-registrations during teardown land in a fresh registry and won't be invoked again in the same pass. New test `test_clear_push_unsubs_safe_when_unsub_reregisters` proves the snapshot invariant.

File: custom_components/lock_code_manager/providers/_base.py [hunk 2]
```diff
@@ -239,6 +249,45 @@ async def _execute_rate_limited(
...
+    @callback
+    def _register_push_unsub(self, unsub: Callable[[], None]) -> None:
+        """
+        Register a push-subscription unsub for base teardown management.
+
+        Providers append integration-specific unsub callables (cluster
+        listeners, event subscriptions, MQTT unsubscribes, bus listeners)
+        here so ``_clear_push_unsubs`` can release them in one place.
+        """
+        self._push_unsubs.append(unsub)
```
[the remark this answers is not in the document]
Comment (raman325) [reply custom_components/lock_code_manager/providers/_base.py:277]: Fixed in b2118f22. Tightened the docstring to explicitly state the scope is the push-subscription lifecycle (`subscribe_push_updates` / `setup_push_subscription`) and to call out that Home Assistant event-bus listeners with the setup/unload lifecycle (like Z-Wave JS's) must NOT be registered here.

**Comment by raman325** [comment]
Pushed `a381add6` addressing the re-review pass:

**Fixed:**
- **Schlage rollback-disconnect blind spot** (was: only the outer delete catch flagged disconnect; a `LockDisconnected` from the inner rollback was logged and silently swallowed). Switched both providers' tracking to a `first_disconnect: LockDisconnected | None` slot so every catch path — including the Schlage rollback — feeds the same summary raise.
- **Lost exception chain.** End-of-loop raise now uses `raise LockDisconnected(...) from first_disconnect` so the original cause (timeout, OSError, etc.) survives in tracebacks.
- **Message rot risk:** "disconnect during *initial* tag pass" wasn't accurate when called from hard-refresh. Softened to "disconnect during tag pass; will be retried on reconnect."
- **`pytest.raises` match=**: existing disconnect tests now anchor on the new summary message text so a regression that removes the summary raise but leaks an unrelated `LockDisconnected` from a deeper layer would fail.
- **New test** `test_async_setup_retries_after_disconnect_on_delete_step` for Schlage covers the delete-catch branch that the prior test missed.

**Deferred** (defensible — not B/E's regressions or scope creep):
- ZHA / Z-Wave broad `except Exception` (recurring concern, would require deliberate exception-type narrowing per provider)
- Reconnect path not driven through `_async_on_integration_loaded` (same structural test-infra gap as the prior idempotency test)
- Sequence-lock concurrency test strengthening (impl correctness not in question)
- ZHA operation-event drop / Z-Wave bare disconnect strings (out of scope)
- Schlage "ghost code" rollback edge case (design limit; orthogonal)
